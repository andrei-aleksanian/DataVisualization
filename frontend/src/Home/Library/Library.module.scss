.index {
  width: 100vw;
  min-height: 100vh;
  background: rgb(2, 0, 36);
  background: linear-gradient(135deg, rgba(2, 0, 36, 1) 0%, rgba(80, 20, 52, 1) 85%);
}

.index h1 {
  text-align: center;
  padding: 2rem 0;
  font-size: 3rem;
  color: var(--color-text);
}

.ExamplesContainer {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.Example {
  width: 350px;
  height: 350px;
  margin: 2rem;
  background-color: #eee;
  border-radius: 1rem;
}

.Example img {
  cursor: pointer;
  width: 350px;
  height: 200px;
  min-height: 200px;
  min-width: 350px;
}

.ExampleInfo {
  padding: 1rem;
}

.ExampleInfo * {
  margin-bottom: 1rem;
}

.Button {
  font-size: 1rem;
  padding: 0.5rem 0.9rem;
  height: auto;
  margin: 0 0 0 -3px;
  border-radius: 0.5rem;
}
