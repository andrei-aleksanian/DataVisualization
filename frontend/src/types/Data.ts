export type Points2D = [number, number, 0];

export type Points3D = [number, number, number];

// Data IN, uncolored:
export interface Data {
  labels: number[];
}

export interface Data2D extends Data {
  points: Points2D[];
}

export interface Data3D extends Data {
  points: Points3D[];
}

// Data here is not used atm
export const DATA: Data2D = {
  points: [
    [0, 0, 0],
    [0, 1, 0],
    [0, 2, 0],
    [3, 1, 0],
    [3, 2, 0],
    [3, 1, 0],
    [-3, 4, 0],
    [-3, 1, 0],
  ],
  labels: [1, 1, 1, 0, 0, 0, 3, 3],
};

// Data WITHIN React, colored:

export interface DataColored {
  colors: string[];
}

export interface Data2DColored extends DataColored {
  points: Points2D[];
}
export interface Data3DColored extends DataColored {
  points: Points3D[];
}
