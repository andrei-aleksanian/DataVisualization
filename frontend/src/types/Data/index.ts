export * from './Data';
export * from './DataPerseverance';
