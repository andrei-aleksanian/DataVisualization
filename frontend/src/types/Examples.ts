export interface ExampleProps {
  name: string;
  id: number;
  imagePath: string;
  description?: string;
}
