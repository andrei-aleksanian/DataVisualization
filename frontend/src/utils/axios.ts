import axios, { AxiosInstance } from 'axios';

const getInstance = (): AxiosInstance => {
  let baseURL = window.location.origin;
  // localhost:3000 is for development only
  // potential todo: change this to an env variable
  if (window.location.origin === 'http://localhost:3000') {
    baseURL = 'http://localhost:8080';
  }

  return axios.create({
    baseURL: `${baseURL}/api`,
  });
};

export default getInstance();
