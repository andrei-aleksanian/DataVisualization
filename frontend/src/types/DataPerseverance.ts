import { DataColored, Points2D, Points3D } from './Data';

// Data for Preservance demo (demo 2)
export interface DataPerseverance extends DataColored {
  prevPartsave: number[];
  prevWrongInLow: number[][];
  prevWrongInHigh: number[][];
}

export interface DataPerseverance2D extends DataPerseverance {
  points: Points2D[];
}
export interface DataPerseverance3D extends DataPerseverance {
  points: Points3D[];
}
