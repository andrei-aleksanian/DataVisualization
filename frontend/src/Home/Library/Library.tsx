import { useEffect, useState } from 'react';
import { LinkBack } from 'Components/Link';
import { ExampleProps } from 'types/Examples';

import { useHistory } from 'react-router-dom';
import getId from 'utils/getId';
import { Popup } from 'Components/UI';
import fetchExamples from './services';

import classes from './Library.module.scss';

export const TEXT_H1 = 'Examples Library';
export const TEXT_POPUP = "Couldn't load examples, please try again later";

const Library = () => {
  const [examples, setExamples] = useState([] as ExampleProps[]);
  const [error, setError] = useState(null as null | string);
  const history = useHistory();

  const Example = ({ name, description, id, imagePath }: ExampleProps) => (
    <div className={classes.Example}>
      <img
        src={`http://localhost:8080/images/${imagePath}`}
        alt={`example ${name} image`}
        onClick={() => history.push(`/examples/${id}`)}
      />
      <div className={classes.ExampleInfo}>
        <h3>{name}</h3>
        <p>{description}</p>
      </div>
    </div>
  );

  useEffect(() => {
    let isActive = true;
    const getExamples = async () => {
      const [data, newError] = await fetchExamples();

      if (!isActive) return;
      if (data) setExamples(data);
      setError(newError);
    };
    getExamples();

    return () => {
      isActive = false;
    };
  }, []);

  return (
    <div className={classes.index}>
      {error && <Popup text={TEXT_POPUP} onClick={() => history.push('/')} />}
      <LinkBack link="/" />
      <h1>{TEXT_H1}</h1>
      <div className={classes.ExamplesContainer}>
        {examples.map((e) => (
          <Example {...e} key={getId('example')} />
        ))}
      </div>
    </div>
  );
};

export default Library;
