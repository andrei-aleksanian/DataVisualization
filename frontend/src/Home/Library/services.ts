import axios from 'utils/axios';
import { ExampleProps } from 'types/Examples';

const fetchExamples = async () => {
  try {
    const { data }: { data: ExampleProps[] } = await axios.get('/examples');
    return [data, null];
  } catch (error) {
    return [null, error];
  }
};

export default fetchExamples;
