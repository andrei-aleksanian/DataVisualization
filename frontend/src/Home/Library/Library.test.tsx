import { render, screen, waitFor } from '@testing-library/react';

import { MemoryRouter } from 'react-router-dom';
import { ExampleProps } from 'types/Examples';
import axios from 'utils/axios';
import Library, { TEXT_POPUP } from '.';

export const placeholderExamples: ExampleProps[] = [
  {
    id: 1,
    name: 'test',
    imagePath: 'test.jpg',
    description: 'description',
  },
];

export const mockAxios = () => {
  return jest.spyOn(axios, 'get');
};

test('Library matches snapshot', async () => {
  const mockedAxios = mockAxios();
  mockedAxios.mockResolvedValueOnce({ data: placeholderExamples });

  const { asFragment } = await waitFor(async () => render(<Library />, { wrapper: MemoryRouter }));
  expect(asFragment()).toMatchSnapshot();
});

describe('Library fetches data correctly', () => {
  test('Library fetches data on render', async () => {
    const mockedAxios = mockAxios();
    mockedAxios.mockResolvedValueOnce({ data: placeholderExamples });

    await waitFor(() => {
      render(<Library />, { wrapper: MemoryRouter });
    });
    const name = await waitFor(() => screen.getByText(placeholderExamples[0].name));
    const descriptoin = await waitFor(() =>
      screen.getByText(placeholderExamples[0].description ? placeholderExamples[0].description : 'description')
    );
    const image = await waitFor(() => screen.getByAltText(`example ${placeholderExamples[0].name} image`));
    const popup = await waitFor(() => screen.queryByText(TEXT_POPUP));

    expect(name).toBeInTheDocument();
    expect(descriptoin).toBeInTheDocument();
    expect(image).toBeInTheDocument();
    expect(popup).not.toBeInTheDocument();
    expect(mockedAxios).toHaveBeenCalledTimes(1);
  });

  test('Library fetches data on render', async () => {
    const mockedAxios = mockAxios();
    mockedAxios.mockRejectedValueOnce(new Error('Something went wrong!'));

    await waitFor(() => {
      render(<Library />, { wrapper: MemoryRouter });
    });

    const popup = await waitFor(() => screen.getByText(TEXT_POPUP));

    expect(popup).toBeInTheDocument();
    expect(mockedAxios).toHaveBeenCalledTimes(1);
  });
});
