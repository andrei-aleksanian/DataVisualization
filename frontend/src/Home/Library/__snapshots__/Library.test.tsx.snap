// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Library matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="index"
  >
    <a
      class="index Back"
      href="/"
    >
      <img
        alt="back"
        src="arrow.svg"
      />
    </a>
    <h1>
      Examples Library
    </h1>
    <div
      class="ExamplesContainer"
    >
      <div
        class="Example"
      >
        <img
          alt="example test image"
          src="localhost/api/images/test.jpg"
        />
        <div
          class="ExampleInfo"
        >
          <h3>
            test
          </h3>
          <p>
            description
          </p>
          <button
            class="Button Button active"
            type="button"
          >
            See Example
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
